microcosm_flask/tests/conventions/test_upload.py

```diff
@@ -167,3 +195,26 @@ def test_upload_for(self):
         assert_that(response_data, is_(equal_to(dict(
             id=str(person_id),
         ))))
+        assert_that(self.controller.calls, contains(
+            has_entries(
+                files=contains(contains("file", anything(), "hello.txt")),
+                extra="something",
+                person_id=person_id,
+            ),
+        ))
+
+    def test_upload_multipart(self):
+        response = self.client.post(
+            "/api/file",
+            data=dict(
+                file=(BytesIO(b("Hello World\n")), "hello.txt"),
+                extra="special",
+            ),
+        )
+        assert_that(response.status_code, is_(equal_to(204)))
+        assert_that(self.controller.calls, contains(
+            has_entries(
+                files=contains(contains("file", anything(), "hello.txt")),
+                extra="special",
```

Comment:
This is the main thing.

Previously, we'd pass non-file parameters in the query string; now we can easily pull them from the form body too.

Reply:
Nice